numpy/core/src/umath/scalarmath.c.src

```diff
@@ -569,6 +571,91 @@ static void
 
 /*** END OF BASIC CODE **/
 
+/**
+ * Find the descriptor for a builtin or Python numerical scalar.
+ * This function should only be used for scalar math fast paths.
+ *
+ * @param obj The object for which to find the descriptor
+ * @param descr The descriptor that was found.
+ * @returns 1 if the other object is a Python type and thus must not
+ *          handle the operation, otherwise 0. -1 _without_ an error set
+ *          if no descriptor was found.
+ */
+static int
+descr_from_basic_scalar(PyObject *obj, PyArray_Descr **descr)
+{
+    /* TODO: We could try giving defined scalars a chance... */
+    int type_num = PyArray_TypeNumFromNumPyScalarType(
+            (PyObject *)Py_TYPE(obj), 0);
+    if (type_num != NPY_NOTYPE) {
+        *descr = PyArray_DescrFromType(type_num);
+        return 0;
+    }
+    else if (PyFloat_CheckExact(obj)) {
+        *descr = PyArray_DescrFromType(NPY_DOUBLE);
+        return 1;
+    }
+    else if (PyLong_CheckExact(obj)) {
+        if (PyLong_AsLong(obj) == -1 && PyErr_Occurred()) {
+            PyErr_Clear();
+            return -1;
+        }
+        *descr = PyArray_DescrFromType(NPY_LONG);
+        return 1;
+    }
+    else if (PyComplex_CheckExact(obj)) {
+        *descr = PyArray_DescrFromType(NPY_CDOUBLE);
+        return 1;
+    }
+    return -1;
+}
+
+
+/**
+ * This function attempts to compare NumPy or Python scalars
+ * The operation is done by getting value of both scalars
+ * and calling richcompare on them.
```

Comment:
Technically, right now it only gets the value of one scalar, and then does a recursive call, right now? Maybe we should even just get the value of both scalars though and do it as it is documented.

Reply:
It got super interesting when I tried this :) :
```
-            if (item_self != NULL) {
-                ret = PyObject_RichCompare(item_self, other, cmp_op);
+            data_other = scalar_value(other, NULL);
+            item_other = other_descr->f->getitem(data_other, NULL);
+            if (item_self != NULL && item_other != NULL) {
+                ret = PyObject_RichCompare(item_self, item_other, cmp_op);
             }
             Py_XDECREF(item_self);
+            Py_XDECREF(item_other);
```

Here is output:
```
>>> np.float16(3) == 3
False
>>> np.int16(3) == 3
False
>>> np.int32(3) == 3
False
>>> np.int64(3) == 3
True
>>> np.int64(3) == 3
True
```

I shall get to the bottom of this

Reply:
Ok I'll circle back to this after the tests, I reached still here and got confused more as to what's happening: https://github.com/python/cpython/blob/c9b8e9c421b57acdcaf24fab0c93bc29b3ef7c67/Objects/object.c#L658